### PR TITLE
os/bluestore: avoid recursive calls in SimpleCollectionListIterator.

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -694,16 +694,14 @@ class SimpleCollectionListIterator : public CollectionListIterator {
   ghobject_t m_oid;
 
   void get_oid() {
-    if (!valid()) {
-      return;
+    m_oid = ghobject_t();
+    while(m_it->valid() && is_extent_shard_key(m_it->key())) {
```
Comment (trociny) [inline src/os/bluestore/BlueStore.cc:698]: nit: a space between `while` and `(` is missing.

**Review by trociny — APPROVED** [review]
LGTM.

Just wondering, did you see or do you expect problems with the recursive version? I just want to understand do we need to backport this.

**Comment by ifed01** [comment]
> LGTM.
> 
> Just wondering, did you see or do you expect problems with the recursive version? I just want to understand do we need to backport this.

Actually this depends on the amount of shards in onode. Theoretically this might hit us one day for some large object.... IMO it's better to prevent beforehand

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-10-03_09:43:24-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-04_04:38:32-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/

**Comment by tchaikov** [comment]
```
2020-10-02T15:16:19.439+0000 7f16b2804540 -1 error: monitor data filesystem reached concerning levels of available storage space (available: 0% 645 MiB)
you may adjust 'mon data avail crit' to a lower value to make this go away (default: 5%)

Using guessed paths /home/jenkins-build/build/workspace/ceph-api/build/lib/ ['/home/jenkins-build/build/workspace/ceph-api/qa', '/home/jenkins-build/build/workspace/ceph-api/build/lib/cython_modules/lib.3', '/home/jenkins-build/build/workspace/ceph-api/src/pybind']
Traceback (most recent call last):
  File "../qa/tasks/vstart_runner.py", line 1596, in <module>
    exec_test()
  File "../qa/tasks/vstart_runner.py", line 1412, in exec_test
    remote.run(args=args, env=vstart_env, timeout=(3 * 60))
  File "../qa/tasks/vstart_runner.py", line 357, in run
    return self._do_run(**kwargs)
  File "../qa/tasks/vstart_runner.py", line 425, in _do_run
    proc.wait()
  File "../qa/tasks/vstart_runner.py", line 208, in wait
    raise CommandFailedError(self.args, self.exitstatus)
teuthology.exceptions.CommandFailedError: Command failed with status 28: ['../src/vstart.sh', '-n', '--nolockdep']
```

**Pull request merged**